### PR TITLE
Add a preflight check that asserts docker connectivity

### LINKED ISSUE #554: Improved error message when Docker is not running
**What happened**:

If you run `kind create cluster` when no Docker daemon is running, kind exits with a unclear message:

```
Error: could not list clusters: failed to list nodes: exit status 1
```

**What you expected to happen**:

A clearer error message, making the experience nicer for the user.

Since the errors are wrapped, and `docker` is executed fairly deep in the codebase, I suggest the following error message:

```
Error: could not list clusters: failed to list nodes, docker might be unavailable: exit status 1
```

**How to reproduce it (as minimally and precisely as possible)**:

1. Don't run a docker daemon
2. Run `kind create cluster` 

**Environment:**

- kind version: (use `kind version`): v0.4.0-alpha (bed5b7ef6c8ceb69606bfa656bd772e0834ff428)

---

I'll submit a PR with a fix for this, if you think that this is a good idea.

### PR DESCRIPTION
**Pull request by zegl**
Assert that the docker daemon is running, and that we're able to connect to it.
This is done by executing 'docker ps', and checking the exit code.

#39
#554

### CODE REVIEW

File: pkg/preflight/runtime/docker.go [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+		if err := check(); err != nil {
+			return errors.Wrap(err, "Preflight check failed")
+		}
+	}
+
+	return nil
+}
+
+// dockerIsRunning asserts that the docker daemon is running and is responsive
+func dockerIsRunning() error {
```
Comment (aojea) [inline pkg/preflight/runtime/docker.go]: I think this should go to `pkg/container/docker`

File: pkg/preflight/runtime/docker.go [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+// Preflight runs checks to make sure that the container runtime
+// is working as expected
+func Preflight() error {
+	checks := []func() error{
+		dockerIsRunning,
+	}
+
+	for _, check := range checks {
+		if err := check(); err != nil {
+			return errors.Wrap(err, "Preflight check failed")
```
Comment (neolit123) [inline pkg/preflight/runtime/docker.go]: error strings should start with lowercase.
[1]

File: pkg/preflight/runtime/docker.go [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+	}
+
+	return nil
+}
+
+// dockerIsRunning asserts that the docker daemon is running and is responsive
+func dockerIsRunning() error {
+	err := exec.Command("docker", "ps").Run()
+	if err != nil {
+		return errors.Wrap(err, "Could not connect to a docker daemon")
```
Comment (neolit123) [inline pkg/preflight/runtime/docker.go]: [1]
`docker` should be upper case.

File: pkg/preflight/runtime/docker.go (hunk 2 above)
Comment (zegl) [reply pkg/preflight/runtime/docker.go]: Thanks, I've fixed it. :)

**Comment by BenTheElder** [comment]
Hi, thanks again for the PR!

If you check the current code, you'll see we actually already do a filtered `docker ps` to list the existing clusters, what we really need is to surface this error better instead of making another call.

I'm going to discuss at today's meeting (in 9 minutes!) and update the preflight checks issue as well to go from pre-flight to post-failure, which I think will help avoid piling on lots of slow checks (not that this one is, but presumably we'd add more) in the normal case, and instead diagnose the environment on failure.

**Comment by BenTheElder** [comment]
going another route, thanks

**Pull request closed without merging**